drivers/can/can_xlnx_zynq.h

```diff
@@ -0,0 +1,278 @@
+
+/* Interrupt status / enable / clear bits: TRM appendix B.5, p. 805 ff. */
+#define CAN_XLNX_ZYNQ_IRQ_TX_EMPTY BIT(14)
+#define CAN_XLNX_ZYNQ_IRQ_TX_WATERMARK BIT(13)
+#define CAN_XLNX_ZYNQ_IRQ_RX_WATERMARK BIT(12)
+#define CAN_XLNX_ZYNQ_IRQ_SLEEP_MODE_EXIT BIT(11)
+#define CAN_XLNX_ZYNQ_IRQ_SLEEP_MODE_ENTER BIT(10)
+#define CAN_XLNX_ZYNQ_IRQ_BUS_OFF BIT(9)
+#define CAN_XLNX_ZYNQ_IRQ_MESSAGE_ERROR BIT(8)
+#define CAN_XLNX_ZYNQ_IRQ_RX_NOT_EMPTY BIT(7)
+#define CAN_XLNX_ZYNQ_IRQ_RX_OVERFLOW BIT(6)
+#define CAN_XLNX_ZYNQ_IRQ_RX_UNDERFLOW BIT(5)
+#define CAN_XLNX_ZYNQ_IRQ_MESSAGE_RX BIT(4)
+#define CAN_XLNX_ZYNQ_IRQ_TXHPB_FULL BIT(3)
+#define CAN_XLNX_ZYNQ_IRQ_TX_FULL BIT(2)
+#define CAN_XLNX_ZYNQ_IRQ_MESSAGE_TX BIT(1)
+#define CAN_XLNX_ZYNQ_IRQ_ARBITRATION_LOST BIT(0)
+
+/* Timestamp Control register bits: TRM appendix B.5, p. 812 f. */
+#define CAN_XLNX_ZYNQ_CTR_CLEAR_TIMESTAMP BIT(0)
+
+/* Watermark Interrupt register offsets and masks: TRM appendix B.5, p. 813 f. */
+#define CAN_XLNX_ZYNQ_WIR_TX_EMPTY_OFFSET 8
+#define CAN_XLNX_ZYNQ_WIR_TX_EMPTY_MASK	0xFF
+#define CAN_XLNX_ZYNQ_WIR_RX_FULL_OFFSET 0
+#define CAN_XLNX_ZYNQ_WIR_RX_FULL_MASK 0xFF
+
+/* FIFO registers offsets and masks: TRM appendix B.5, p. 814 ff. */
+#define CAN_XLNX_ZYNQ_FIFO_IDR_IDH_OFFSET 21
+#define CAN_XLNX_ZYNQ_FIFO_IDR_IDH_MASK 0x7FF
+#define CAN_XLNX_ZYNQ_FIFO_IDR_SRRRTR BIT(20)
+#define CAN_XLNX_ZYNQ_FIFO_IDR_IDE BIT(19)
+#define CAN_XLNX_ZYNQ_FIFO_IDR_IDL_OFFSET 1
+#define CAN_XLNX_ZYNQ_FIFO_IDR_IDL_MASK 0x3FFFF
+#define CAN_XLNX_ZYNQ_FIFO_IDR_RTR BIT(0)
+#define CAN_XLNX_ZYNQ_FIFO_DLCR_DLC_OFFSET 28
+#define CAN_XLNX_ZYNQ_FIFO_DLCR_DLC_MASK 0xF
+#define CAN_XLNX_ZYNQ_FIFO_DLCR_RXT_MASK 0xFFFFF
+
+/* I/O pin control macros, depending on this feature being enabled */
+#ifdef CONFIG_PINCTRL
+#define CAN_XLNX_ZYNQ_DEV_PINCTRL_DEFINE(inst) PINCTRL_DT_INST_DEFINE(inst);
+#define CAN_XLNX_ZYNQ_DEV_PINCTRL_INIT(inst) .pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),
+#else
+#define CAN_XLNX_ZYNQ_DEV_PINCTRL_DEFINE(inst)
+#define CAN_XLNX_ZYNQ_DEV_PINCTRL_INIT(inst)
+#endif /* CONFIG_PINCTRL */
+
+/* Mode/state transition validation retry count */
+#define CAN_XLNX_ZYNQ_MODE_STATE_CHANGE_RETRIES 8192
+
+/* Per-instance device initialization macros */
+#define CAN_XLNX_ZYNQ_DEV_DATA(inst)\
+static struct can_xlnx_zynq_dev_data can_xlnx_zynq_##inst##_data = {\
+	.common = {0},\
+	.base = 0x0,\
+	.state = CAN_STATE_STOPPED,\
+	.tx_errors = 0,\
+	.rx_errors = 0,\
+	.tx_callback = NULL,\
+	.tx_user_data = NULL,\
+	.timing = {0} \
+};
+
+#define CAN_XLNX_ZYNQ_DEV_CONFIG(inst)\
+static const struct can_xlnx_zynq_dev_cfg can_xlnx_zynq_##inst##_cfg = {\
+	.common = CAN_DT_DRIVER_CONFIG_INST_GET(inst, 0, 1000000),\
+	DEVICE_MMIO_NAMED_ROM_INIT(reg_base, DT_DRV_INST(inst)),\
+	.irq_config_func = can_xlnx_zynq_##inst##_irq_config,\
+	.irq = DT_INST_IRQN(inst),\
+	CAN_XLNX_ZYNQ_DEV_PINCTRL_INIT(inst)\
+	.clock_frequency = DT_INST_PROP_OR(inst, clock_frequency, 0),\
```

Comment:
Same comment as last time you submitted this: https://github.com/zephyrproject-rtos/zephyr/pull/45101#discussion_r962361342

Let's get https://github.com/zephyrproject-rtos/zephyr/pull/64996 in first.

Reply:
I fully agree that having clock control first is the cleaner way - for the Zynq, that is. Assuming that this PR would likely go through first, I didn't add any #ifdef'd frequency get calls for the baudrate calculation as of yet, I could still add that in order to make it clearer that those facilities should be used whenever available. 

However, the loophole of also allowing the old fashioned way (shouldn't have made clock-frequency mandatory though) and using DT_INST_PROP_OR(clock_frequency, 0) as the macro that acquires the data from the DT is aimed at the UltraScale. I did clean up some of the technical debt that in particular the soc code layer for this chip has and I'm hoping to add it step by step once I'm back from vacation, however, I don't see s clock control driver on the horizon for the UltraScale yet...

I can either throw the legacy way out completely, thus going Zynq-only for now and blocking on the clock control driver, or I could fully implement the use of clock control when available and making it clearer that the legacy way shouldn't be used if not running on the UltraScale.